File: rust/bron_kerbosch/src/bron_kerbosch1o.rs
//! Naive Bron-Kerbosch algorithm, optimized

use graph::{connected_vertices, UndirectedGraph, Vertex, VertexSetLike};
use pile::Pile;
use reporter::Reporter;

type Clique<'a> = Pile<'a, Vertex>;

pub fn explore<VertexSet, Graph, Rprtr>(graph: &Graph, reporter: &mut Rprtr)
where
    VertexSet: VertexSetLike,
    Graph: UndirectedGraph<VertexSet = VertexSet>,
    Rprtr: Reporter,
{
    let candidates = connected_vertices(graph);
    let num_candidates = candidates.len();
    if num_candidates > 0 {
        visit(
            graph,
            reporter,
            candidates,
            VertexSet::with_capacity(num_candidates),
            Pile::new(),
        );
    }
}

fn visit<VertexSet, Graph, Rprtr>(
    graph: &Graph,
    reporter: &mut Rprtr,
    mut candidates: VertexSet,
    mut excluded: VertexSet,
    clique: Clique,
) where
    VertexSet: VertexSetLike,
    Graph: UndirectedGraph<VertexSet = VertexSet>,
    Rprtr: Reporter,
{
    debug_assert!(!candidates.is_empty());
    debug_assert!(candidates.all(|&v| graph.degree(v) > 0));
    debug_assert!(excluded.all(|&v| graph.degree(v) > 0));
    debug_assert!(candidates.is_disjoint(&excluded));

    loop {
        let v = candidates.pop_arbitrary().unwrap();
        let neighbours = graph.neighbours(v);
        let neighbouring_candidates: VertexSet = neighbours.intersection_collect(&candidates);
        if !neighbouring_candidates.is_empty() {
            let neighbouring_excluded: VertexSet = neighbours.intersection_collect(&excluded);
            visit(
                graph,
                reporter,
                neighbouring_candidates,
                neighbouring_excluded,
                clique.place(v),
            );
        } else {
            if neighbours.is_disjoint(&excluded) {
                reporter.record(clique.place(v).collect());
            }
            if candidates.is_empty() {
                break;
            }
        }
        excluded.insert(v);
    }
}

File: python3/bron_kerbosch1o.py
# coding: utf-8

from graph import UndirectedGraph, Vertex
from reporter import Reporter
from typing import List, Set


def bron_kerbosch1o(graph: UndirectedGraph, reporter: Reporter):
    '''Naive Bron-Kerbosch algorithm optimized'''
    if candidates := graph.connected_vertices():
        visit(
            graph=graph,
            reporter=reporter,
            candidates=candidates,
            excluded=set(),
            clique=[])


def visit(graph: UndirectedGraph, reporter: Reporter, candidates: Set[Vertex],
          excluded: Set[Vertex], clique: List[Vertex]):
    assert candidates

    while candidates:
        v = candidates.pop()
        neighbours = graph.adjacencies[v]
        assert neighbours
        neighbouring_candidates = candidates.intersection(neighbours)
        if neighbouring_candidates:
            visit(
                graph=graph,
                reporter=reporter,
                candidates=neighbouring_candidates,
                excluded=excluded.intersection(neighbours),
                clique=clique + [v])
        else:
            if excluded.isdisjoint(neighbours):
                reporter.record(clique + [v])
        excluded.add(v)
